src/categorical_algebra/ACSetViews.jl

```diff
@@ -0,0 +1,185 @@
+""" Data structure for views of attributed C-sets
+"""
+module ACSetViews
+export ACSetView, backing, @compute_prop, @select_where
+
+using ...CSetDataStructures
+using ..CSets
+using ..FinSets
+using ...Theories: CatDesc, AttrDesc, dom, codom_num
+
+using MLStyle: @match
+using PrettyTables: pretty_table
+using Tables
+
+# Data types
+############
+
+""" One useful way of thinking of ACSetViews is that they are
+dataframes backed by ACSets.
+
+They are also useful for computing limits
+
+P is a symbol, the object in the schema that the view focuses onto
+"""
+struct ACSetView{A <: ACSet, P, Attrs <: NamedTuple} <: AbstractArray{Attrs,1}
+  backing :: A
+  parts :: AbstractVector{Int}
+  function ACSetView(backing::A, P::Symbol, parts::AbstractVector{Int}) where
+      {CD <: CatDesc,AD <: AttrDesc{CD},Ts <: Tuple, A <: ACSet{CD,AD,Ts}}
+    attr_names = filter(a -> dom(AD,a) == P, AD.attr)
+    attr_types = map(a -> Ts.parameters[codom_num(AD,a)], attr_names)
+    row_type = NamedTuple{attr_names, Tuple{attr_types...}}
+    new{A,P,row_type}(backing,parts)
+  end
+  function ACSetView(backing::A, P::Symbol) where
+      {CD <: CatDesc,AD <: AttrDesc{CD},Ts <: Tuple, A <: ACSet{CD,AD,Ts}}
+    ACSetView(backing,P,1:nparts(backing,P))
+  end
+end
+
+# We're going to overload getproperty, so we use these instead
+
+backing(av::ACSetView) = getfield(av,:backing)
+
+CSetDataStructures.parts(av::ACSetView) = getfield(av,:parts)
+
+
+# Overloaded accessors
+######################
+
+Base.size(av::ACSetView) = size(parts(av))
+
+@generated function Base.getindex(av::ACSetView{<:ACSet,P,Attrs}, i::Int) where
+    {P, Attrs <: NamedTuple}
+  tuple_args = map(Attrs.parameters[1]) do attr
+    :($attr = av[i,$(Expr(:quote, attr))])
+  end
+  Expr(:tuple,tuple_args...)
+end
+
+function Base.getindex(av::ACSetView{<:ACSet,P,<:NamedTuple},i::Int,name::Union{Symbol,Vector{Symbol}}) where {P}
+  subpart(backing(av), parts(av)[i], name)
+end
+
+function Base.getindex(av::ACSetView{<:ACSet,P,<:NamedTuple},i,name::Union{Symbol,Vector{Symbol}}) where {P}
+  subpart(backing(av), view(parts(av),i), name)
+end
+
+function Base.getindex(av::ACSetView{<:ACSet,P,<:NamedTuple},name::Union{Symbol,Vector{Symbol}}) where {P}
+  subpart(backing(av), parts(av), name)
+end
+
+function Base.getindex(av::ACSetView{<:ACSet,P,<:NamedTuple},indicator::BitArray{1}) where {P}
+  ACSetView(backing(av),P,parts(av)[indicator])
+end
+
+Base.getproperty(av::ACSetView, name::Symbol) =
+  subpart(backing(av), parts(av), name)
+
+function Base.show(io::IO, ::MIME"text/plain", av::T) where {ACS <: AbstractACSet, P, T<:ACSetView{ACS,P}}
+  print(io, "ACSetView of object $P")
+  println(io)
+  table = tables(backing(av))[P][parts(av)]
+  if !(eltype(table) <: NamedTuple{(),Tuple{}} || isempty(table))
+    pretty_table(io, table, nosubheader=true,
+                 show_row_number=true, row_number_column_title=string(P))
+  end
+end
+
+# Tables.jl interface
```

Comment:
It looks like most of the Tables.jl interface is not covered by tests. Would be good to add a simple test for each function.